src/hash_value.cpp

```diff
@@ -21,13 +21,36 @@ bool HashValue::compare(const void *a, const void *b, void *env) {
     return a_p->key.send((Env *)env, SymbolValue::intern("eql?"), { b_p->key })->is_truthy();
 }
 
+ValuePtr HashValue::compare_by_identity(Env *env) {
+    assert_not_frozen(env);
+    this->m_is_comparing_by_identity = true;
+    this->rehash(env);
+    return this;
+}
+
+ValuePtr HashValue::is_comparing_by_identity() {
+    if (m_is_comparing_by_identity) {
+        return TrueValue::the();
+    } else {
+        return FalseValue::the();
+    }
+}
+
 ValuePtr HashValue::get(Env *env, ValuePtr key) {
     Key key_container;
     key_container.key = key;
-    key_container.hash = key.send(env, SymbolValue::intern("hash"))->as_integer()->to_nat_int_t();
+    key_container.hash = generate_key_hash(env, key);
     return m_hashmap.get(&key_container, env);
 }
 
+nat_int_t HashValue::generate_key_hash(Env *env, ValuePtr key) {
+    if (m_is_comparing_by_identity) {
+        return key->object_id();
```

Comment:
This won't return a value that is evenly distributed, so we'll have to do something like:

```suggestion
        return TM::Hashmap<void*>::hash_ptr(key);
```

Reply:
I'm not sure if it's the same but the suggestion you did would not compile.. I changed it to 
```cpp
return TM::Hashmap<void*>::hash_ptr(key.value());
```
Is this fine?

I guess it won't change the behavior of the rehash spec that is failing though, because the spec does not use `compare_by_identity`.

Reply:
Yes, that is correct. I typed it without checking.

I'll take a look at the failing spec with a debugger a little later!

Reply:
Great! I definitely have to get into this debugging thing myself at some point...